pkg/dyninst/decode/marshal.go

```diff
@@ -0,0 +1,473 @@
+	if err := writeTokens(enc,
+		jsontext.BeginObject,
+		jsontext.String("function"),
+		jsontext.String(sl.Function),
+		jsontext.String("file"),
+		jsontext.String(sl.File),
+		jsontext.String("line"),
+		jsontext.Int(int64(sl.Line)),
+		jsontext.EndObject,
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Decoder) encodeValue(
+	enc *jsontext.Encoder,
+	dataItems map[typeAndAddr]output.DataItem,
+	currentlyEncoding map[typeAndAddr]struct{},
+	irType ir.Type,
+	data []byte,
+) error {
+	switch v := irType.(type) {
+	case *ir.BaseType:
+		return d.encodeBaseTypeValue(enc, v, data)
+	case *ir.PointerType:
+		if len(data) < int(v.GetByteSize()) {
+			return errors.New("passed data not long enough for pointer")
+		}
+		addr := binary.NativeEndian.Uint64(data)
+		key := typeAndAddr{
+			irType: uint32(v.Pointee.GetID()),
+			addr:   addr,
+		}
+		var address uint64
+		pointedValue, dataItemExists := dataItems[key]
+		if !dataItemExists {
+			address = key.addr
+		} else {
+			address = pointedValue.Header().Address
+		}
+		var err error
+
+		if err = writeTokens(enc,
+			jsontext.BeginObject,
+			jsontext.String("address"),
+			jsontext.Int(int64(address)),
+		); err != nil {
+			return err
+		}
+		if _, alreadyEncoding := currentlyEncoding[key]; !alreadyEncoding && dataItemExists {
+			currentlyEncoding[key] = struct{}{}
+			defer delete(currentlyEncoding, key)
+			if err = writeTokens(enc, jsontext.String("value")); err != nil {
+				return err
+			}
+			if err = d.encodeValue(enc,
+				dataItems,
+				currentlyEncoding,
+				d.program.Types[ir.TypeID(pointedValue.Header().Type)],
+				pointedValue.Data(),
+			); err != nil {
+				return fmt.Errorf("could not encode referenced value: %w", err)
+			}
+		}
+		if err = writeTokens(enc, jsontext.EndObject); err != nil {
+			return err
+		}
+	case *ir.StructureType:
+		var err error
+		if err = writeTokens(enc, jsontext.BeginObject); err != nil {
+			return err
+		}
+		structure := irType.(*ir.StructureType)
+		for _, field := range structure.Fields {
+			if err = writeTokens(enc, jsontext.String(field.Name)); err != nil {
+				return err
+			}
+			if err = d.encodeValue(enc,
+				dataItems,
+				currentlyEncoding,
+				field.Type,
+				data[field.Offset:field.Offset+field.Type.GetByteSize()],
+			); err != nil {
+				return err
+			}
+		}
+		if err = writeTokens(enc, jsontext.EndObject); err != nil {
+			return err
+		}
+
+	case *ir.ArrayType:
+		var err error
+		if err = writeTokens(enc, jsontext.BeginArray); err != nil {
+			return err
+		}
+
+		elementType := v.Element
+		elementSize := int(elementType.GetByteSize())
+		numElements := int(v.Count)
+		for i := range numElements {
+			elementData := data[i*elementSize : (i+1)*elementSize]
+			if err := d.encodeValue(enc,
+				dataItems, currentlyEncoding, v.Element, elementData); err != nil {
+				return err
+			}
+		}
+		if err = writeTokens(enc, jsontext.EndArray); err != nil {
+			return err
+		}
+		return nil
```

Comment:
why is it that some of these switch cases return and some fall to the end of the function which has a return nil? Let's remove the `return nil` from the end of the function and force each case to return explicitly.